### ISSUE TITLE
fastJoin and context.params.provider

### ISSUE THREAD

**Issue by sxwebdev** [feathers-plus/feathers-hooks-common#398]
Hey. I have a trouble when I use fastJoin hook in my after hooks.

In this example:
![image](https://user-images.githubusercontent.com/3129246/40873300-6f39fab4-6666-11e8-9788-f726ab6c580f.png)
![image](https://user-images.githubusercontent.com/3129246/40873303-832ae1d2-6666-11e8-8b92-c29eb56d4abc.png)

This operation works great, and all data fetch from organizations service. But in my organization after hooks a have this code

![image](https://user-images.githubusercontent.com/3129246/40873320-d812cb56-6666-11e8-8b58-2568aaed1cd3.png)

The problem is `hook.params.provider = undefined` and `discard` not works. Clients get not protected data.

I try pass provider from context in user after hooks. But it did not help.

How to resolve this? Thanks

**Comment by bertho-zero** [comment]
You should add the provider to the find method call in your join resolver.
```js
user.organization = (await context.app
  .service("organizations")
  .find({
    query: { _id: user.organization },
    paginate: false,
    provider: context.provider
  }))[0];
```

**Comment by sxwebdev** [comment]
> You should add the provider to the find method call in your join resolver.

I did it.

![image](https://user-images.githubusercontent.com/3129246/40880400-a12a693a-66b8-11e8-9ed5-8f8121be8470.png)

and this does not work

![image](https://user-images.githubusercontent.com/3129246/40880392-8984e0a8-66b8-11e8-9a19-c4f5d374b590.png)

protected data return to client

**Comment by sxwebdev** [comment]
If I console hook provider in organization service

![image](https://user-images.githubusercontent.com/3129246/40880449-ccffc734-66b9-11e8-914c-fa7d3c5e3c64.png)

I get undefined

![image](https://user-images.githubusercontent.com/3129246/40880453-dc109348-66b9-11e8-9a79-33e9c49b0f31.png)

**Comment by bertho-zero** [comment]
You should move your `when` hook on the service `organizations`, or discard `organization.*`

**Comment by sxwebdev** [comment]
`You should move your when hook on the service organizations, or discard organization.*`

like this?

![image](https://user-images.githubusercontent.com/3129246/40902296-35d78bb6-67dc-11e8-809d-bf37e09e0828.png)

this not helped

**Comment by sxwebdev** [comment]
> If you expected your service call's context.provider !== undefined but it is, is the service call being made within another hook?

Hey. Thanks for feedback.

I pass context  to after hook in users service, and call this from client. Like this.

![image](https://user-images.githubusercontent.com/3129246/40917464-0bec7192-680c-11e8-953c-9844a22cda0b.png)

Then client context pass to find method by fastJoin.

![image](https://user-images.githubusercontent.com/3129246/40917576-6cb0a0ca-680c-11e8-86dc-932966f651d4.png)

> That is normal when the service call is made on the server. Its not undefined only when the service call is made by a client.

Organizations hooks must define client `context.provider`. But this is not so.

**Comment by eddyystop** [comment]
Why do you have `...context` in `(...context) => async (user, context) =>` ?

**Comment by sxwebdev** [comment]
This was not a problem, but I updated the code as you described above

![image](https://user-images.githubusercontent.com/3129246/40923872-96dfb716-681e-11e8-9eee-7f20bf587865.png)

and `after hook`

![image](https://user-images.githubusercontent.com/3129246/40923959-c5c4af82-681e-11e8-9fa2-0a96ffb637fb.png)

This did not help

**Comment by sxwebdev** [comment]
> You can share your code like this

ok) I did not know how to do this

**Comment by sxwebdev** [comment]
> You can use the makeCallingParams utility function for this https://feathers-plus.github.io/v1/feathers-hooks-common/index.html#makecallingparams

```
const joinOrg = {
	joins: {
		organization: () => async (user,context) =>  {
			user.organization = (await context.app.service('organizations').find(
				makeCallingParams(context,{_id:user.organization},'user',{paginate:false,_populate: false})
			))[0]
		}
	}
}
```

This did not help. Clients receives unprotected data.

**Comment by eddyystop** [comment]
You have said `fastJoin` works as expected. So the issue is the `when`.

Write a custom hook that logs `context.provider`. Add it as the first and last hook in `all`. Also add it between every hook in `all`.

If `context.provider` is `undefined` at the start, then you rproblem is elsewhere.

**Comment by bertho-zero** [comment]
You need to discard `organization.dadata` or add the discard hook directly on the organization service.

**Comment by sxwebdev** [comment]
> Write a custom hook that logs context.provider. Add it as the first and last hook in all. Also add it between every hook in all.

ok. I got a strange sequence:

![image](https://user-images.githubusercontent.com/3129246/40935610-4e0299aa-6841-11e8-91d5-89fb2f00a220.png)

**Comment by sxwebdev** [comment]
> You need to discard organization.dadata or add the discard hook directly on the organization service.

If I completely remove `dadata`, it will not be available on the server or on the client.

**Comment by sxwebdev** [comment]
I do not understand why it is not here

![image](https://user-images.githubusercontent.com/3129246/40938949-fb169358-684b-11e8-842f-240dc2c0f4a6.png)

**Comment by sxwebdev** [comment]
If I don`t use fastJoin and use this example. All works expected

```js
get: [
		async hook => {
			hook.result.organization = (await hook.app.service('organizations').find({
				query:{_id:hook.result.organization},
				paginate:false,
				provider:hook.params.provider}
			))[0]
			return hook;
		}
	]
```

**Comment by sxwebdev** [comment]
```js
alterItems( (u,hook) => {
	u.organization = (async () => await hook.app.service('organizations').get(u.organization,hook))();
})
```
and this example also does not work

**Comment by eddyystop** [comment]
Please stop providing fragments of code. Provide the whole hook module. We can't guess what you are doing wrong from fragments.

I also asked the following:
```text
Why do you have ...context in (...context) => async (user, context) => ? I suggest you use different names for every context variable. Your problem may be that one context is "shadowing" another context and your code is not using the context you expect.

I'm also confused why you are using fastJoin(context => joinOrg) rather than just fastJoin(joinOrg). The latter seems to do the same as the former.
```

Have you done that?

**Comment by sxwebdev** [comment]
Of course. I did it and showed you in this post [https://github.com/feathers-plus/feathers-hooks-common/issues/398#issuecomment-394380052](url)

**Comment by eddyystop** [comment]
Once again, past the whole module here.

**Comment by sxwebdev** [comment]
users hooks
```js
const { authenticate } = require('@feathersjs/authentication').hooks;
const verifyHooks = require('feathers-authentication-management').hooks;
const { iff,isProvider,preventChanges,disallow,when,discard, fastJoin, alterItems } = require('feathers-hooks-common');
const accountService = require('../auth-management/notifier');
const isEnabled = require('../../hooks/is-enabled');

const {
  hashPassword, protect
} = require('@feathersjs/authentication-local').hooks;

const joinOrg = {
	joins: {
		organization: () => async (user,context) =>  {
			user.organization = (await context.app.service('organizations').find(
				makeCallingParams(context,{_id:user.organization},'user',{paginate:false,_populate: false})
			))[0]
		}
	}
}

module.exports = {
  before: {
    all: [],
    find: [ authenticate('jwt'), isEnabled() ],
    get: [ authenticate('jwt'), isEnabled() ],
    create: [
    	hashPassword(),
    	when(
    		hook => !hook.data.verifyPhone && !hook.data.verifyPhoneToken && !hook.data.checkData,
    		verifyHooks.addVerification()
    	)
    ],
    update: [ disallow('external') ],
    patch: [ authenticate('jwt'),

    	iff(isProvider('external'), preventChanges(true,
	      'isVerified',
	      'verifyToken',
	      'verifyShortToken',
	      'verifyExpires',
	      'verifyChanges',
	      'resetToken',
	      'resetShortToken',
	      'resetExpires'
	    )),
    ],
    remove: [ authenticate('jwt') ]
  },

  after: {
    all: [
		fastJoin(joinOrg),
		protect('password'),
		when(
			hook => hook.params.provider,
			discard('password', '_computed', 'verifyExpires', 'resetExpires', 'verifyChanges', 'verifyToken', 'verifyShortToken')
		)
	],
	find: [],
	get: [],
	create: [
		async hook => {
			if (!hook.params.provider) return hook;
			const user = hook.result;

			if(hook.data && hook.data.email && user && !hook.data.verifyPhone && !hook.data.verifyPhoneToken && !hook.data.checkData) {
				accountService(hook.app).notifier('resendVerifySignup', hook.result);
				return hook;
			}
		},
		verifyHooks.removeVerification()
    ],
    update: [],
    patch: [],
    remove: []
  },

  error: {
    all: [],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  }
};
```

organization hooks

```js
const { authenticate } = require('@feathersjs/authentication').hooks;
const hooks = require('feathers-authentication-hooks');
const { discard,when,fastJoin, disallow, alterItems } = require('feathers-hooks-common');

const restrictToRoles = [
	hooks.restrictToRoles({
			roles: ['director'],
			fieldName: 'position'
		})
	];

const joinContacts = {
	joins: {
		contacts: () => async (org,context) => org.contacts = await context.app.service('organization-contacts').find({
			query: { organization: org._id },
			paginate: false
		})
	}
}

module.exports = {
  before: {
    all: [ authenticate('jwt') ],
    find: [],
    get: [],
    create: [
    	hooks.associateCurrentUser()
    ],
    update: [ disallow('external') ],
    patch: [
    	async hook => {
    		hook.data.updatedAt = Date.now();
    	}
    ],
    remove: [ ...restrictToRoles ]
  },

  after: {
    all: [
    	when(
			hook => hook.params.provider,
			discard('kfocus', 'atiData', 'userId', 'dadata', 'verifyChanges', 'adminEmail', 'telegram', 'type')
		),
		alterItems( o => { o.management = !o.management ? {} : o.management } ),
		fastJoin(joinContacts),
    ],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  },

  error: {
    all: [],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  }
};
```

**Comment by eddyystop** [comment]
Initial comment is `makeCallingParams` is not defined in users hooks module.

**Comment by sxwebdev** [comment]
> Initial comment is makeCallingParams is not defined in users hooks module.

Excuse me. I updated code. I reduced the code for testing and forgot to edit this line.

**Comment by eddyystop** [comment]
Could you run the fastJoin hook twice, one right after the other, and see if that makes a difference? Thanks.

**Comment by sxwebdev** [comment]
I tried it like you said. I Launched fastJoin hook twice and it did not help
```js
all: [
		fastJoin(joinOrg),
		protect('password'),
		when(
			hook => hook.params.provider,
			discard('password', '_computed', 'verifyExpires', 'resetExpires', 'verifyChanges', 'verifyToken', 'verifyShortToken')
		),
		fastJoin(joinOrg)
]
```

**Comment by eddyystop** [comment]
If I understand correctly, you may be saying context.params.provider may not exist when context.params.connection exists. Is this correct?

That would be affect so many hooks.

**Comment by eddyystop** [comment]
Frankly, sorry about alterItems. I forgot to merge it quite a while ago.

**Comment by sxwebdev** [comment]
> If I understand correctly, you may be saying context.params.provider may not exist when context.params.connection exists. Is this correct?

When a client makes a request to the server via rest or socketio, `context.params.connection` contains the correct value of `provider`. `context.params.provider` contains an invalid value and = undefined.

**Comment by eddyystop** [comment]
This is a big bug. Could you please post an issue in @feathers/feathers ?

**Comment by daffl** [comment]
It might be easier to track down with a complete example but I am not able to reproduce this issue (https://github.com/feathersjs/feathers/issues/884).

**Comment by eddyystop** [comment]
Such a bug would have caused issues with other hooks and I'd expect them to have been reporetd by now.

**Comment by daffl** [comment]
Agreed. It is possible to accidentally remove or reset `context.params.provider` but without a minimal breaking example I don't think we can reproduce what's happening.

**Comment by eddyystop** [comment]
I'm closing this issue since an example has yet to be provided. Please open another issue with the provided example as this one is getting too long.

Thanks.
